x/avs/keeper/keeper.go

```diff
@@ -44,47 +46,83 @@ func (k Keeper) Logger(ctx sdk.Context) log.Logger {
 }
 
 func (k Keeper) AVSInfoUpdate(ctx sdk.Context, params *AVSRegisterOrDeregisterParams) error {
+	avsInfo, _ := k.GetAVSInfo(ctx, params.AvsAddress)
+
+	action := params.Action
+
+	if action == RegisterAction && avsInfo != nil {
+		return errorsmod.Wrap(types.ErrAlreadyRegistered, fmt.Sprintf("the error input arg is:%s", params.AvsAddress))
+	}
+
+	avs := &types.AVSInfo{
+		Name:               params.AvsName,
+		AvsAddress:         params.AvsAddress,
+		SlashAddr:          params.SlashContractAddr,
+		AvsOwnerAddress:    params.AvsOwnerAddress,
+		AssetId:            params.AssetID,
+		AvsUnbondingEpochs: uint32(params.MinimumDelegation),
+		MinimumDelegation:  sdk.NewIntFromUint64(params.UnbondingEpochs),
+		AvsEpoch:           nil,
+		OperatorAddress:    nil,
+	}
+	if action == RegisterAction && avsInfo == nil {
+		return k.SetAVSInfo(ctx, avs)
+	}
+
+	if avsInfo == nil {
+		return errorsmod.Wrap(types.ErrUnregisterNonExistent, fmt.Sprintf("the error input arg is:%s", avsInfo))
+	}
+
+	//TODO:if avs DeRegisterAction check UnbondingEpochs
+	// if avsInfo.Info.AvsUnbondingEpochs < currenUnbondingEpoch - regUnbondingEpoch {
+	//	return errorsmod.Wrap(err, fmt.Sprintf("not qualified to deregister %s", avsInfo))
+	//}
+	return k.DeleteAVSInfo(ctx, params.AvsAddress)
+}
+
+func (k Keeper) AVSInfoUpdateWithOperator(ctx sdk.Context, params *OperatorOptParams) error {
 	operatorAddress := params.OperatorAddress
 	opAccAddr, err := sdk.AccAddressFromBech32(operatorAddress)
 	if err != nil {
 		return errorsmod.Wrap(err, fmt.Sprintf("error occurred when parse acc address from Bech32,the addr is:%s", operatorAddress))
 	}
 	if !k.operatorKeeper.IsOperator(ctx, opAccAddr) {
-		return errorsmod.Wrap(delegationtypes.ErrOperatorNotExist, "AVSInfoUpdate: invalid operator address")
-	}
-
-	action := params.Action
-
-	if action == RegisterAction {
-		return k.SetAVSInfo(ctx, params.AvsName, params.AvsAddress, operatorAddress, params.AssetID)
+		return errorsmod.Wrap(delegationtypes.ErrOperatorNotExist, fmt.Sprintf("AVSInfoUpdate: invalid operator address:%s", operatorAddress))
 	}
 	avsInfo, err := k.GetAVSInfo(ctx, params.AvsAddress)
-	if err != nil {
+	if err != nil || avsInfo.GetInfo() == nil {
```

Comment:
there are two lines call **avsInfo.GetInfo()**, can we call once ?

Reply:
fixed